File: app/src/main/java/com/notai/ui/viewmodel/NoteViewModel.java
package com.notai.ui.viewmodel;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.notai.NoteAIApplication;
import com.notai.data.model.Note;
import com.notai.data.model.NoteWithTags;
import com.notai.data.repository.NoteRepository;

import java.util.ArrayList;
import java.util.List;

/**
 * 笔记 ViewModel
 */
public class NoteViewModel extends ViewModel {
    
    private static final String TAG = "NoteViewModel";
    
    private final NoteRepository noteRepository;
    private final MutableLiveData<String> searchQuery = new MutableLiveData<>("");
    private final LiveData<List<NoteWithTags>> notes;
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    
    public NoteViewModel(NoteRepository noteRepository) {
        this.noteRepository = noteRepository;
        
        try {
            // 使用 Transformations.switchMap 根据搜索查询切换数据源
            notes = Transformations.switchMap(searchQuery, query -> {
                if (query == null || query.trim().isEmpty()) {
                    return noteRepository.getAllNotesWithTags();
                } else {
                    return noteRepository.searchNotesWithTags(query);
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "Error initializing NoteViewModel", e);
            throw e;
        }
    }
    
    public LiveData<List<NoteWithTags>> getNotes() {
        return notes;
    }
    
    public LiveData<String> getSearchQuery() {
        return searchQuery;
    }
    
    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }
    
    public void searchNotes(String query) {
        searchQuery.setValue(query != null ? query : "");
    }
    
    public void insertNote(Note note, List<Long> tagIds) {
        isLoading.postValue(true);
        noteRepository.insertNote(note, tagIds != null ? tagIds : new ArrayList<>(), noteId -> {
            isLoading.postValue(false);
        });
    }
    
    public void updateNote(Note note, List<Long> tagIds) {
        isLoading.postValue(true);
        note.setUpdatedAt(System.currentTimeMillis());
        noteRepository.updateNote(note, tagIds != null ? tagIds : new ArrayList<>());
        // 更新操作是异步的，但由于没有回调，我们延迟一点再设置为 false
        // 实际上，由于 Room 的 LiveData 会自动更新，我们可以立即设置为 false
        isLoading.postValue(false);
    }
    
    public void deleteNote(long noteId) {
        isLoading.postValue(true);
        noteRepository.deleteNote(noteId);
        // 删除操作是异步的，但由于没有回调，我们延迟一点再设置为 false
        // 实际上，由于 Room 的 LiveData 会自动更新，我们可以立即设置为 false
        isLoading.postValue(false);
    }
    
    public LiveData<NoteWithTags> getNoteWithTagsById(long noteId) {
        return noteRepository.getNoteWithTagsById(noteId);
    }
    
    public static class Factory implements ViewModelProvider.Factory {
        @Override
        @SuppressWarnings("unchecked")
        public <T extends ViewModel> T create(Class<T> modelClass) {
            if (modelClass.isAssignableFrom(NoteViewModel.class)) {
                try {
                    NoteAIApplication application = NoteAIApplication.getInstance();
                    return (T) new NoteViewModel(application.getNoteRepository());
                } catch (Exception e) {
                    Log.e(TAG, "Error creating NoteViewModel", e);
                    throw new RuntimeException("Failed to create NoteViewModel", e);
                }
            }
            throw new IllegalArgumentException("Unknown ViewModel class");
        }
    }
}

File: app/src/main/java/com/notai/ui/viewmodel/NoteViewModel.kt
package com.notai.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.notai.NoteAIApplication
import com.notai.data.model.Note
import com.notai.data.model.NoteWithTags
import com.notai.data.repository.NoteRepository
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * 笔记 ViewModel
 */
class NoteViewModel(
    private val noteRepository: NoteRepository
) : ViewModel() {
    
    private val _notes = MutableStateFlow<List<NoteWithTags>>(emptyList())
    val notes: StateFlow<List<NoteWithTags>> = _notes.asStateFlow()
    
    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private var searchJob: Job? = null
    
    init {
        loadNotes()
    }
    
    fun loadNotes() {
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            noteRepository.getAllNotesWithTags().collect { notesList ->
                _notes.value = notesList
            }
        }
    }
    
    fun searchNotes(query: String) {
        _searchQuery.value = query
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            if (query.isBlank()) {
                noteRepository.getAllNotesWithTags().collect { notesList ->
                    _notes.value = notesList
                }
            } else {
                noteRepository.searchNotesWithTags(query).collect { notesList ->
                    _notes.value = notesList
                }
            }
        }
    }
    
    fun insertNote(note: Note, tagIds: List<Long> = emptyList()) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                noteRepository.insertNote(note, tagIds)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun updateNote(note: Note, tagIds: List<Long> = emptyList()) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val updatedNote = note.copy(updatedAt = System.currentTimeMillis())
                noteRepository.updateNote(updatedNote, tagIds)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun deleteNote(noteId: Long) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                noteRepository.deleteNote(noteId)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    suspend fun getNoteWithTagsById(noteId: Long): NoteWithTags? {
        return noteRepository.getNoteWithTagsById(noteId)
    }
    
    companion object {
        fun provideFactory(): ViewModelProvider.Factory {
            return object : ViewModelProvider.Factory {
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T {
                    val application = NoteAIApplication.getInstance()
                    return NoteViewModel(application.noteRepository) as T
                }
            }
        }
    }
}

